pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,566 @@
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sPodName:   pod.Name,
+			logfields.K8sNamespace: pod.Namespace,
+		}).Error("failed to get valid pod metadata")
+		return
+	}
+
+	if policies, ok := manager.policyPods[id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			manager.deleteEgressMap(config, podData)
+		}
+		delete(manager.policyPods, id)
+	}
+	delete(manager.podDataStore, id)
+}
+
+func (manager *Manager) getLocalPodsForNamespace(ns *slim_corev1.Namespace) []*slim_corev1.Pod {
+	var retPods []*slim_corev1.Pod
+
+	podStore := manager.storeGetter.GetStore("pod")
+	for _, podItem := range podStore.List() {
+		pod, ok := podItem.(*slim_corev1.Pod)
+		if !ok {
+			continue
+		}
+
+		if pod.Namespace != ns.Name {
+			continue
+		} else {
+			retPods = append(retPods, pod)
+		}
+	}
+
+	return retPods
+}
+
+func (manager *Manager) OnUpdateNamespace(ns *slim_corev1.Namespace) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	// Ns labels have been updated, need to re-evaluate all pods within ns.
+	pods := manager.getLocalPodsForNamespace(ns)
+	if len(pods) == 0 {
+		return
+	}
+	for _, pod := range pods {
+		manager.OnUpdatePodLocked(pod, true /* removeOld */, true /* upsertNew */)
+	}
+}
+
+func (manager *Manager) getPodMetadata(pod *slim_corev1.Pod, podIPs []string) (*podMetadata, error) {
+	podId := types.NamespacedName{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	namespace, err := manager.getNamespaceForPod(pod)
+	if namespace == nil || err != nil {
+		log.Warn("Failed in getting namespace")
+		data := &podMetadata{
+			ips:    podIPs,
+			labels: pod.GetLabels(),
+			id:     podId,
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (manager *Manager) getNamespaceForPod(pod *slim_corev1.Pod) (*slim_corev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := manager.storeGetter.GetStore("namespace")
+	nsItem, exists, err := nsStore.GetByKey(pod.GetNamespace())
+	if err != nil {
+		log.WithError(err).Warn("Failed to get namespace")
+		return nil, err
+	}
+	if !exists {
+		log.WithFields(logrus.Fields{"nsName": nsName}).Warn("Namespace does not exist")
+		return nil, nil
+	}
+	return nsItem.(*slim_corev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	for _, policyConfig := range manager.policyConfigs {
+		if policyConfig.egressIP.String() == config.egressIP.String() {
+			return fmt.Errorf(
+				"CiliumEgressNatPolicy for egress ip %v already exists, %v is not valid config.",
+				config.egressIP.String(), config.id.Name)
```

Comment:
```suggestion
			return fmt.Errorf("CiliumEgressNatPolicy for egress IP %v already exists", config.egressIP.String())
```

This should just state the reason why it's not a valid config. The caller then should format the error returned and say:

```
err := isValidConfig()
if err != nil {
  return fmt.Error("invalid config: %w", err)
}
```

This should result in an error string that looks like:

```
invalid config: CiliumEgressNatPolicy for egress IP x.x.x.x already exists
```

Reply:
this is fixed.